internal/ledger/compare.go

```diff
@@ -0,0 +1,420 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package ledger
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hyperledger/fabric/common/ledger/util"
+	"github.com/hyperledger/fabric/core/ledger/kvledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/privacyenabledstate"
+	"github.com/hyperledger/fabric/internal/fileutil"
+	"github.com/pkg/errors"
+)
+
+// Compare - Compares two ledger snapshots and outputs the differences in snapshot records
+// This function will overwrite the file at outputPath if it already exists
+func Compare(snapshotDir1 string, snapshotDir2 string, outputFile string) (count int, err error) {
+	count = 0
+
+	// Check the hashes between two files
+	hashPath1 := filepath.Join(snapshotDir1, kvledger.SnapshotSignableMetadataFileName)
+	hashPath2 := filepath.Join(snapshotDir2, kvledger.SnapshotSignableMetadataFileName)
+
+	equal, err := snapshotsComparable(hashPath1, hashPath2)
+	if err != nil {
+		return 0, err
+	}
+
+	if equal {
+		return 0, errors.New("both snapshots public state hashes are same. Aborting compare")
+	}
+
+	// Create the output file
+	jsonOutputFile, err := newJSONFileWriter(outputFile)
+	if err != nil {
+		return 0, err
+	}
+
+	// Create snapshot readers to read both snapshots
+	snapshotReader1, err := privacyenabledstate.NewSnapshotReader(snapshotDir1,
+		privacyenabledstate.PubStateDataFileName, privacyenabledstate.PubStateMetadataFileName)
+	if err != nil {
+		return 0, err
+	}
+	snapshotReader2, err := privacyenabledstate.NewSnapshotReader(snapshotDir2,
+		privacyenabledstate.PubStateDataFileName, privacyenabledstate.PubStateMetadataFileName)
+	if err != nil {
+		return 0, err
+	}
+
+	// Read each snapshot record  to begin looking for divergences
+	namespace1, snapshotRecord1, err := snapshotReader1.Next()
+	if err != nil {
+		return 0, err
+	}
+	namespace2, snapshotRecord2, err := snapshotReader2.Next()
+	if err != nil {
+		return 0, err
+	}
+
+	// Main snapshot record comparison loop
+	for snapshotRecord1 != nil && snapshotRecord2 != nil {
+
+		// nsKeys used for comparing snapshot records
+		key1 := &nsKey{namespace: namespace1, key: snapshotRecord1.Key}
+		key2 := &nsKey{namespace: namespace2, key: snapshotRecord2.Key}
+
+		// Determine the difference in records by comparing nsKeys
+		switch nsKeyCompare(key1, key2) {
+
+		case 0: // Keys are the same, look for a difference in records
+			if !(proto.Equal(snapshotRecord1, snapshotRecord2)) {
+				// Keys are the same but records are different
+				diffRecord, err := newDiffRecord(namespace1, snapshotRecord1, snapshotRecord2)
+				if err != nil {
+					return 0, err
+				}
+				// Add difference to output JSON file
+				err = jsonOutputFile.addRecord(*diffRecord)
+				if err != nil {
+					return 0, err
+				}
+				count++
+
```

Comment:
Instead of keeping track of count++ at various places, you can keep track of this inside `jsonOutputFile`